File: 0802-find-eventual-safe-states/0802-find-eventual-safe-states.cpp
class Solution {
public:
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        queue<int> q;
        vector<int> indegree(graph.size(),0);
        vector<int> adj[graph.size()];
        for(int i=0;i<graph.size();i++){
             for(auto &x:graph[i]){
                adj[x].push_back(i);
                indegree[i]++;
            }
            if(indegree[i]==0){
               q.push(i);
           }
        }
        vector<int> ans;
        while(!q.empty()){
            auto it = q.front();
            ans.push_back(it);
            q.pop();
            for(auto j:adj[it]){
                indegree[j]--;
                if(indegree[j]==0){
                    q.push(j);
                }
            }
        }
        sort(ans.begin(),ans.end());
        return ans;
    }
};

File: 0802-find-eventual-safe-states/0802-find-eventual-safe-states.java
class Solution {
    public List<Integer> eventualSafeNodes(int[][] graph) {
        List<List<Integer>> adjMat = new ArrayList<>();
        int[] outDegree = new int[graph.length];
        for(int i=0;i<graph.length;i++){
            adjMat.add(new ArrayList<>());
        }
        Queue<Integer> q = new LinkedList<>();
        for(int i=0;i<graph.length;i++){
            for(int a: graph[i]){
                adjMat.get(a).add(i);
                outDegree[i]++;
            }
            if(outDegree[i]==0){
                q.offer(i);
            }
        }
        List<Integer> ans = new ArrayList<>();
        while(!q.isEmpty()){
            Integer ele = q.poll();
            ans.add(ele);
            for(int a: adjMat.get(ele)){
                outDegree[a]--;
                if(outDegree[a]==0){
                    q.offer(a);
                }
            }
        }
        Collections.sort(ans);
        return ans;
    }
}